src/bokeh/plotting/_legends.py

```diff
@@ -67,7 +67,12 @@ def _find_legend_item(label, legend):
     return None
 
 def _get_or_create_legend(plot):
-    legends = plot.select(type=Legend)
+    # Using the simpler plot.select(type=Legend) to find the existing legend
+    # here is very inefficient on already populated plots, therefore we do it
+    # like this:
+    panels = plot.above + plot.below + plot.left + plot.right + plot.center
+    legends = [obj for obj in panels if isinstance(obj, Legend)]
```

Comment:
This won't work in bokeh 3.2, where it will be possible to create nested layouts with legends. Being able to traverse plot's children is important, it just has to be limited in scope. Maybe let's apply `.select()` to `panels`, flatten and filter the result.

Reply:
Ah, thanks for that hint! So you mean like this, basically?

    legends = sum([list(panel.select({'type': Legend})) for panel in panels], [])
 
Unfortunately, that `select` suffers from the same problem as the current code. Is there any other future-proof approach that doesn't involve iterating over _all_ objects with `select`?

Reply:
We would need to implement a customized traversal, but for that we have to know the interface for container models, which we don't have. I suppose this the existing fix is fine for now. I would add a `TODO` comment regarding traversing the children in the future.

Reply:
If you can do that and add some tests for `_get_or_create_legend()` in near future, then we can consider this 3.1.

Reply:
I've added the `TODO` as well as a some unit tests :+1: